File: android/src/main/java/com/supermap/smNative/collector/SMMediaCollector.java
package com.supermap.smNative.collector;

import android.content.Context;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;

import com.supermap.RNUtils.DateUtil;
import com.supermap.RNUtils.LocationTransfer;
import com.supermap.RNUtils.MediaUtil;
import com.supermap.component.MapWrapView;
import com.supermap.data.Color;
import com.supermap.data.CoordSysTransMethod;
import com.supermap.data.CoordSysTransParameter;
import com.supermap.data.CoordSysTranslator;
import com.supermap.data.CursorType;
import com.supermap.data.DatasetVector;
import com.supermap.data.FieldType;
import com.supermap.data.GeoLine;
import com.supermap.data.GeoLineM;
import com.supermap.data.GeoStyle;
import com.supermap.data.Point2D;
import com.supermap.data.Point2Ds;
import com.supermap.data.PrjCoordSys;
import com.supermap.data.PrjCoordSysType;
import com.supermap.data.QueryParameter;
import com.supermap.data.Recordset;
import com.supermap.data.Size2D;
import com.supermap.interfaces.mapping.SMap;
import com.supermap.mapping.Layer;
import com.supermap.mapping.Map;
import com.supermap.mapping.MapControl;
import com.supermap.smNative.SMLayer;
import com.supermap.smNative.components.InfoCallout;

import java.lang.reflect.Array;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

/**
 * @Author: shanglongyang
 * Date:        2019/5/28
 * project:     iTablet
 * package:     iTablet
 * class:
 * description:
 */
public class SMMediaCollector {
    private static SMMediaCollector instance;
    private static String mediaPath;
    public static String sdcard = android.os.Environment.getExternalStorageDirectory().getAbsolutePath();

    public SMMediaCollector() {
//        this.context = context;
    }

    public static SMMediaCollector getInstance() {
        if (instance == null) {
            instance = new SMMediaCollector();
        }
        return instance;
    }

    public static String getMediaPath() {
        return mediaPath;
    }

    public static void setMediaPath(String mediaPath) {
        SMMediaCollector.mediaPath = mediaPath;
    }

    public static boolean hasMediaData(Layer layer) {
        DatasetVector dsVector = (DatasetVector)layer.getDataset();

        int tag1 = dsVector.getFieldInfos().indexOf("MediaName");
        int tag2 = dsVector.getFieldInfos().indexOf("ModifiedDate");
        int tag3 = dsVector.getFieldInfos().indexOf("MediaFilePaths");
        int tag4 = dsVector.getFieldInfos().indexOf("Description");
        int tag5 = dsVector.getFieldInfos().indexOf("HttpAddress");

        if(tag1==-1 || tag2==-1 || tag3==-1 || tag4==-1 || tag5==-1)
            return false;
        if (dsVector.getFieldInfos().get(tag1).getType() != FieldType.TEXT)
            return false;
        if (dsVector.getFieldInfos().get(tag2).getType() != FieldType.TEXT)
            return false;
        if (dsVector.getFieldInfos().get(tag3).getType() != FieldType.TEXT)
            return false;
        if (dsVector.getFieldInfos().get(tag4).getType() != FieldType.TEXT)
            return false;
        if (dsVector.getFieldInfos().get(tag5).getType() != FieldType.TEXT)
            return false;
//        if (dsVector.getPrjCoordSys().getType() != PrjCoordSysType.PCS_EARTH_LONGITUDE_LATITUDE)
//            return false;

        return true;
    }

    public static SMMedia findMediaByLayer(Layer layer, int geoID) {
        SMMedia smMedia = new SMMedia();

        DatasetVector dsVector = (DatasetVector)layer.getDataset();
        Recordset recordset;

        if (!hasMediaData(layer)) {
            return null;
        }

        String filter = "SmID=" + geoID;
        QueryParameter parameter = new QueryParameter();
        parameter.setAttributeFilter(filter);
        parameter.setCursorType(CursorType.STATIC);
        recordset = dsVector.query(parameter);

        smMedia.setDatasourse(layer.getDataset().getDatasource());
        smMedia.setDataset(layer.getDataset());
        smMedia.setFileName((String)recordset.getFieldValue("MediaName"));

        String paths = (String)recordset.getFieldValue("MediaFilePaths");
        if (paths == null) paths = "";
        ArrayList<String> pathArr = new ArrayList<>();
        if (paths.indexOf(",") > 0) {
            pathArr = new ArrayList<>(Arrays.asList(paths.split(",")));
        } else {
            pathArr.add(paths);
        }
        smMedia.setPaths(pathArr);

        double x = recordset.getGeometry().getInnerPoint().getX();
        double y = recordset.getGeometry().getInnerPoint().getY();
        smMedia.setLocation(new Point2D(x, y));

        recordset.dispose();

        return smMedia;
    }

    public static ArrayList<SMMedia>findMediasByLayer(Layer layer) {
        ArrayList<SMMedia> medias = new ArrayList<>();

        DatasetVector datasetVector = (DatasetVector)layer.getDataset();

        if (!hasMediaData(layer)) {
            return null;
        }

        Recordset recordset = datasetVector.getRecordset(false, CursorType.STATIC);
        recordset.moveFirst();

        while (!recordset.isEOF()) {
            SMMedia media = new SMMedia();

            media.setDatasourse(layer.getDataset().getDatasource());
            media.setDataset(layer.getDataset());
            media.setFileName((String)recordset.getFieldValue("MediaName"));

            String paths = (String)recordset.getFieldValue("MediaFilePaths");
            ArrayList<String> pathArr = (ArrayList<String>)Arrays.asList(paths.split(","));
            media.setPaths(pathArr);

            double x = recordset.getGeometry().getInnerPoint().getX();
            double y = recordset.getGeometry().getInnerPoint().getY();
            media.setLocation(new Point2D(x, y));

            medias.add(media);

            recordset.moveNext();
        }

        recordset.dispose();

        return medias;
    }

    public static void addMedias(Context context, Layer layer, View.OnTouchListener listener) {
        if (hasMediaData(layer)) {
            DatasetVector datasetVector = (DatasetVector)layer.getDataset();
            Recordset recordset = datasetVector.getRecordset(false, CursorType.STATIC);
            recordset.moveFirst();
            while (!recordset.isEOF()) {
                SMMedia media = new SMMedia();

                media.setDatasourse(layer.getDataset().getDatasource());
                media.setDataset(layer.getDataset());
                media.setFileName((String)recordset.getFieldValue("MediaName"));

                String paths = (String)recordset.getFieldValue("MediaFilePaths");
                ArrayList<String> pathArr = new ArrayList<>(Arrays.asList(paths.split(",")));
                media.setPaths(pathArr);

                double x = recordset.getGeometry().getInnerPoint().getX();
                double y = recordset.getGeometry().getInnerPoint().getY();
                media.setLocation(new Point2D(x, y));

                String imgPath = sdcard + media.getPaths().get(0);
                InfoCallout callout = SMLayer.addCallOutWithLongitude(context, x, y, imgPath);
                callout.setMediaName(media.getFileName());
                callout.setMediaFilePaths(media.getPaths());
                callout.setLayerName(layer.getName());
                callout.setHttpAddress("");
                callout.setDescription("");

                Date date = new Date();
                callout.setModifiedDate(DateUtil.formatDateToString(date, "yyyy-MM-dd HH:mm:ss"));
                callout.setGeoID(Integer.parseInt(recordset.getFieldValue("SmID").toString()));

                if (listener != null) {
                    callout.setOnTouchListener(listener);
                }

                recordset.moveNext();
            }

            recordset.dispose();
        }
    }

    public static void showMediasByLayer(Context context, Layer layer, View.OnTouchListener listener) {
        if (hasMediaData(layer)) {
            DatasetVector datasetVector = (DatasetVector)layer.getDataset();
            Recordset recordset = datasetVector.getRecordset(false, CursorType.STATIC);
            recordset.moveFirst();
            ArrayList<InfoCallout> callouts = new ArrayList<>();
            while (!recordset.isEOF()) {
                String paths = (String)recordset.getFieldValue("MediaFilePaths");
                String fileName = (String)recordset.getFieldValue("MediaName");
                if (paths == null && (fileName == null || fileName.equals("TourLine"))) {
                    recordset.moveNext();
                    continue;
                }
                SMMedia media = new SMMedia();

                media.setDatasourse(layer.getDataset().getDatasource());
                media.setDataset(layer.getDataset());
                media.setFileName(fileName);
                ArrayList<String> pathArr;
                if (paths != null) {
                    pathArr = new ArrayList<>(Arrays.asList(paths.split(",")));
                } else {
                    pathArr = new ArrayList<>();
                }
                media.setPaths(pathArr);

                double x = recordset.getGeometry().getInnerPoint().getX();
                double y = recordset.getGeometry().getInnerPoint().getY();
                media.setLocation(new Point2D(x, y));

                InfoCallout callout = getCalloutByMedia(context, media, recordset, layer.getName(), listener);

                callouts.add(callout);
                recordset.moveNext();
            }

            addCallouts(callouts);

            recordset.dispose();
        }
    }

    public static InfoCallout getCalloutByMedia(Context context, SMMedia media, Recordset recordset, String layerName, View.OnTouchListener listener) {
        int imgSize = 60;

        Point2D pt = new Point2D(media.getLocation().getX(), media.getLocation().getY());

        final InfoCallout callout = new InfoCallout(context);

        String imagePath = sdcard + media.getPaths().get(0);
        String extension = imagePath.substring(imagePath.lastIndexOf(".") + 1);

        final ImageView img = new ImageView(context);
        LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(imgSize, imgSize);
        img.setLayoutParams(layoutParams);
        img.setMinimumHeight(imgSize);
        img.setMinimumWidth(imgSize);

        if (extension.equals("mp4")) {
            img.setImageBitmap(MediaUtil.getScreenShotImageFromVideoPath(imagePath, imgSize, imgSize));
        } else {
            img.setImageBitmap(MediaUtil.getLocalBitmap(imagePath, imgSize, imgSize));
        }
        callout.setContentView(img);
        callout.setLocation(pt.getX(), pt.getY());

        callout.setMediaName(media.getFileName());
        callout.setMediaFilePaths(media.getPaths());
        callout.setLayerName(layerName);
        callout.setHttpAddress("");
        callout.setDescription("");


        Date date = new Date();
        callout.setModifiedDate(DateUtil.formatDateToString(date, "yyyy-MM-dd HH:mm:ss"));
        callout.setGeoID(Integer.parseInt(recordset.getFieldValue("SmID").toString()));

        if (listener != null) {
            callout.setOnTouchListener(listener);
        }

        return callout;
    }

    public static InfoCallout createCalloutByMedia(Context context, SMMedia media, Recordset recordset, String layerName, View.OnTouchListener listener) {
        int imgSize = 60;

        Point2D pt = new Point2D(media.getLocation().getX(), media.getLocation().getY());

        final InfoCallout callout = new InfoCallout(context);

        String imagePath = sdcard + media.getPaths().get(0);
        String extension = imagePath.substring(imagePath.lastIndexOf(".") + 1);

        final ImageView img = new ImageView(context);
        LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(imgSize, imgSize);
        img.setLayoutParams(layoutParams);
        img.setMinimumHeight(imgSize);
        img.setMinimumWidth(imgSize);

        if (extension.equals("mp4")) {
            img.setImageBitmap(MediaUtil.getScreenShotImageFromVideoPath(imagePath, imgSize, imgSize));
        } else {
            img.setImageBitmap(MediaUtil.getLocalBitmap(imagePath, imgSize, imgSize));
        }
        callout.setContentView(img);
        callout.setLocation(pt.getX(), pt.getY());

        callout.setMediaName(media.getFileName());
        callout.setMediaFilePaths(media.getPaths());
        callout.setLayerName(layerName);
        callout.setHttpAddress("");
        callout.setDescription("");

        Date date = new Date();
        callout.setModifiedDate(DateUtil.formatDateToString(date, "yyyy-MM-dd HH:mm:ss"));
        callout.setGeoID(Integer.parseInt(recordset.getFieldValue("SmID").toString()));

        String paths = "";
        for (int i = 0; i < callout.getMediaFilePaths().size(); i++) {
            paths += callout.getMediaFilePaths().get(i) + (i == callout.getMediaFilePaths().size() - 1 ? "" : ",");
        }

        recordset.edit();
        recordset.setFieldValue("ModifiedDate", callout.getModifiedDate());
        recordset.setFieldValue("MediaFilePaths", paths);
        recordset.setFieldValue("Description", callout.getDescription());
        recordset.setFieldValue("HttpAddress", callout.getHttpAddress());
        recordset.update();

        if (listener != null) {
            callout.setOnTouchListener(listener);
        }

        return callout;
    }

    public static void addCallouts(final ArrayList<InfoCallout> callouts) {
        final SMap sMap = SMap.getInstance();
        sMap.getActivity().runOnUiThread(new Runnable(){
            @Override
            public void run(){
                Map map = SMap.getInstance().getSmMapWC().getMapControl().getMap();
                MapWrapView mapWrapView = (MapWrapView)map.getMapView();
                for (int i = 0; i < callouts.size(); i++) {
                    InfoCallout callout = callouts.get(i);
                    mapWrapView.addCallout(callout, callout.getID());
                }
                map.setCenter(new Point2D(callouts.get(0).getLocationX(), callouts.get(0).getLocationY()));
                if (map.getScale() < 0.000011947150294723098) {
                    map.setScale(0.000011947150294723098);
                }
                map.refresh();

                mapWrapView.showCallOut();
            }
        });
    }

    public static InfoCallout addCalloutByMedia(Context context, SMMedia media, Recordset recordset, String layerName, View.OnTouchListener listener) {
        Point2D pt = new Point2D(media.getLocation().getX(), media.getLocation().getY());

        String imgPath = sdcard + media.getPaths().get(0);
        InfoCallout callout = SMLayer.addCallOutWithLongitude(context, pt.getX(), pt.getY(), imgPath);
        callout.setMediaName(media.getFileName());
        callout.setMediaFilePaths(media.getPaths());
        callout.setLayerName(layerName);
        callout.setHttpAddress("");
        callout.setDescription("");

        Date date = new Date();
        callout.setModifiedDate(DateUtil.formatDateToString(date, "yyyy-MM-dd HH:mm:ss"));
        callout.setGeoID(Integer.parseInt(recordset.getFieldValue("SmID").toString()));

        String paths = "";
        for (int i = 0; i < callout.getMediaFilePaths().size(); i++) {
            paths += callout.getMediaFilePaths().get(i) + (i == callout.getMediaFilePaths().size() - 1 ? "" : ",");
        }

        recordset.edit();
        recordset.setFieldValue("ModifiedDate", callout.getModifiedDate());
        recordset.setFieldValue("MediaFilePaths", paths);
        recordset.setFieldValue("Description", callout.getDescription());
        recordset.setFieldValue("HttpAddress", callout.getHttpAddress());
        recordset.update();

        if (listener != null) {
            callout.setOnTouchListener(listener);
        }
        return callout;
    }

    public static void addLineByMedias(ArrayList<SMMedia> medias, DatasetVector dataset) {
        Point2Ds mPoints = new Point2Ds();
        for (int i = 0; i < medias.size(); i++) {
            Point2D pt = new Point2D(medias.get(i).getLocation().getX(), medias.get(i).getLocation().getY());
            mPoints.add(pt);
        }

        GeoLine mLine = new GeoLine(mPoints);
        GeoStyle style = new GeoStyle();
        style.setMarkerSize(new Size2D(2, 2));
        style.setLineColor(new Color(70, 128, 233));
        style.setLineWidth(1);
        style.setMarkerSymbolID(351);
        style.setFillForeColor(new Color(70, 128, 233));
        mLine.setStyle(style);

        Recordset mRecordset = dataset.getRecordset(false, CursorType.DYNAMIC);

        mRecordset.addNew(mLine);
        mRecordset.moveLast();
        if (mRecordset.edit()) {
            mRecordset.setString("MediaName", "TourLine");
        }

        mRecordset.update();
        mRecordset.dispose();
    }
}

File: ios/native/collector/SMMediaCollector.m
//
//  SMMediaCollector.m
//  Supermap
//
//  Created by Shanglong Yang on 2019/5/24.
//  Copyright © 2019 Facebook. All rights reserved.
//

#import "SMMediaCollector.h"

static SMMediaCollector* sMediaCollector = nil;

@implementation SMMediaCollector

+ (instancetype)singletonInstance{
    
    static dispatch_once_t once;
    
    dispatch_once(&once, ^{
        sMediaCollector = [[self alloc] init];
    });
    
    return sMediaCollector;
}

+ (BOOL)hasMediaData:(Layer *)layer {
    DatasetVector* dsVector = (DatasetVector *)layer.dataset;
    int tag1 = (int)[dsVector.fieldInfos indexOfWithFieldName:@"MediaName"];
    //        int tag2 = (int)[dsVector.fieldInfos indexOfWithFieldName:@"MediaFileType"];
    int tag3 = (int)[dsVector.fieldInfos indexOfWithFieldName:@"ModifiedDate"];
    int tag4 = (int)[dsVector.fieldInfos indexOfWithFieldName:@"MediaFilePaths"];
    int tag5 = (int)[dsVector.fieldInfos indexOfWithFieldName:@"Description"];
    int tag6 = (int)[dsVector.fieldInfos indexOfWithFieldName:@"HttpAddress"];
    
    if(tag1==-1 || tag3==-1 || tag4==-1 || tag5==-1 || tag6==-1)
        return NO;
    if([dsVector.fieldInfos get:tag1].fieldType!=FT_TEXT)
        return NO;
    //        if([dsVector.fieldInfos get:tag2].fieldType!=FT_INT16)
    //            return NO;
    if([dsVector.fieldInfos get:tag3].fieldType!=FT_TEXT)
        return NO;
    if([dsVector.fieldInfos get:tag4].fieldType!=FT_TEXT)
        return NO;
    if([dsVector.fieldInfos get:tag5].fieldType!=FT_TEXT)
        return NO;
    if([dsVector.fieldInfos get:tag6].fieldType!=FT_TEXT)
        return NO;
//    if(dsVector.prjCoordSys.type != PCST_EARTH_LONGITUDE_LATITUDE)
//        return NO;
    
    return YES;
}

+ (SMMedia *)findMediaByLayer:(Layer *)layer geoID:(int)geoID {
    SMMedia* media = [[SMMedia alloc] init];
    
    DatasetVector* dsVector = (DatasetVector *)layer.dataset;
    Recordset* recordset;
    
    if (![SMMediaCollector hasMediaData:layer]) {
        return nil;
    }
    
    NSString* filter = [NSString stringWithFormat:@"SmID=%d", geoID];
    QueryParameter* queryParams = [[QueryParameter alloc] init];
    [queryParams setAttriButeFilter:filter];
    [queryParams setCursorType:STATIC];
    recordset = [dsVector query:queryParams];
    
    media.datasourse = layer.dataset.datasource;
    media.dataset = layer.dataset;
    media.fileName = (NSString *)[recordset getFieldValueWithString:@"MediaName"];
    media.httpAddress = (NSString *)[recordset getFieldValueWithString:@"HttpAddress"];
    media.Description = (NSString *)[recordset getFieldValueWithString:@"Description"];
    
    NSString* paths = (NSString *)[recordset getFieldValueWithString:@"MediaFilePaths"];
    media.paths = [paths componentsSeparatedByString:@","];
    
//    if ([media.paths count] > 0) {
//        media.mediaDicPath = [media.paths[0] stringByDeletingLastPathComponent];
//    }
    
    double x =  [recordset.geometry getInnerPoint].x;
    double y =  [recordset.geometry getInnerPoint].y;
    
    
    media.location = [[Point2D alloc] initWithX:x Y:y];
    
    [recordset dispose];
    
    return media;
}

+ (NSArray *)findMediasByLayer:(Layer *)layer {
    NSMutableArray* medias = [[NSMutableArray alloc] init];
    
    DatasetVector* dsVector = (DatasetVector *)layer.dataset;
    Recordset* recordset;
    
    if (![SMMediaCollector hasMediaData:layer]) {
        return nil;
    }
    
    recordset = [dsVector recordset:NO cursorType:STATIC];
    
    [recordset moveFirst];
    
    while (![recordset isEOF]) {
        SMMedia* media = [[SMMedia alloc] init];
        
        media.datasourse = layer.dataset.datasource;
        media.dataset = layer.dataset;
        media.fileName = (NSString *)[recordset getFieldValueWithString:@"MediaName"];
        media.httpAddress = (NSString *)[recordset getFieldValueWithString:@"HttpAddress"];
        media.Description = (NSString *)[recordset getFieldValueWithString:@"Description"];
        
        NSString* paths = (NSString *)[recordset getFieldValueWithString:@"MediaFilePaths"];
        media.paths = [paths componentsSeparatedByString:@","];
        
        double x =  [recordset.geometry getInnerPoint].x;
        double y =  [recordset.geometry getInnerPoint].y;
        media.location = [[Point2D alloc] initWithX:x Y:y];
        
        [medias addObject:media];
        
        [recordset moveNext];
    }
    
    [recordset dispose];
    
    return medias;
}

+ (void)addMedias:(Layer *)layer gesture:(UITapGestureRecognizer *)gesture {
    DatasetVector* dsVector = (DatasetVector *)layer.dataset;
    Recordset* recordset;
    
    if ([SMMediaCollector hasMediaData:layer]) {
        recordset = [dsVector recordset:NO cursorType:STATIC];
        
        [recordset moveFirst];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            int recordsetIndex = 0;
            while (![recordset isEOF]) {
                SMMedia* media = [[SMMedia alloc] init];
                
                media.datasourse = layer.dataset.datasource;
                media.dataset = layer.dataset;
                media.fileName = (NSString *)[recordset getFieldValueWithString:@"MediaName"];
                media.httpAddress = (NSString *)[recordset getFieldValueWithString:@"HttpAddress"];
                media.Description = (NSString *)[recordset getFieldValueWithString:@"Description"];
                
                NSString* paths = (NSString *)[recordset getFieldValueWithString:@"MediaFilePaths"];
                media.paths = [paths componentsSeparatedByString:@","];
                
                double x =  [recordset.geometry getInnerPoint].x;
                double y =  [recordset.geometry getInnerPoint].y;
                media.location = [[Point2D alloc] initWithX:x Y:y];
                
                NSString* imgPath = [NSString stringWithFormat:@"%@/%@", [NSHomeDirectory() stringByAppendingString:@"/Documents"], media.paths[0]];
                
                InfoCallout* callout = [SMLayer addCallOutWithLongitude:x latitude:y image:imgPath];
                callout.mediaName = media.fileName;
                callout.mediaFilePaths = media.paths;
                //            callout.type = media.mediaType;
                callout.layerName = layer.name;
                callout.httpAddress = @"";
                callout.description = @"";
                NSDate* date = [NSDate date];
                
                NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
                [dateFormatter setDateFormat:@"yyyy-MM-dd HH:mm:ss"];
                callout.modifiedDate = [dateFormatter stringFromDate:date];
                
                callout.geoID = ((NSNumber *)[recordset getFieldValueWithString:@"SmID"]).intValue;
                
                if (gesture) {
                    [callout addGestureRecognizer:gesture];
                }
                
                if ([recordset moveNext]) {
                    recordsetIndex++;
                }
            }
            
            [recordset dispose];
        });
    }
}

+ (void)showMediasByLayer:(Layer *)layer gesture:(UITapGestureRecognizer *)gesture {
    DatasetVector* dsVector = (DatasetVector *)layer.dataset;
    Recordset* recordset;
    
    if ([SMMediaCollector hasMediaData:layer]) {
        recordset = [dsVector recordset:NO cursorType:STATIC];
        
        [recordset moveFirst];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            SMap* sMap = [SMap singletonInstance];
            int recordsetIndex = 0;
            Point2D* center = nil;
            while (![recordset isEOF]) {
                SMMedia* media = [[SMMedia alloc] init];
                
                media.datasourse = layer.dataset.datasource;
                media.dataset = layer.dataset;
                media.fileName = (NSString *)[recordset getFieldValueWithString:@"MediaName"];
                media.httpAddress = (NSString *)[recordset getFieldValueWithString:@"HttpAddress"];
                media.Description = (NSString *)[recordset getFieldValueWithString:@"Description"];
                
                NSString* paths = (NSString *)[recordset getFieldValueWithString:@"MediaFilePaths"];
                if (paths && paths.length > 0) {
                    media.paths = [paths componentsSeparatedByString:@","];
                    
                    double x =  [recordset.geometry getInnerPoint].x;
                    double y =  [recordset.geometry getInnerPoint].y;
                    media.location = [[Point2D alloc] initWithX:x Y:y];
                    
                    InfoCallout* callout = [SMMediaCollector getCalloutByMedia:media recordset:recordset layerName:layer.name segesturelector:gesture];
                    
                    [callout showAt:media.location Tag:callout.layerName];
                    
                    if (!center) center = callout.getLocation;
                }
                
                if ([recordset moveNext]) {
                    recordsetIndex++;
                }
            }
            
            [recordset dispose];
            
            sMap.smMapWC.mapControl.map.center = center;
            if (sMap.smMapWC.mapControl.map.scale < 0.000011947150294723098) {
                sMap.smMapWC.mapControl.map.scale = 0.000011947150294723098;
            }
            [sMap.smMapWC.mapControl.map refresh];
        });
    }
}

+ (InfoCallout *)getCalloutByMedia:(SMMedia *)media recordset:(Recordset *)recordset layerName:(NSString *)layerName segesturelector:(UITapGestureRecognizer *)gesture {
    SMap* sMap = [SMap singletonInstance];
//    Point2D* pt = [[Point2D alloc] initWithX:media.location.x Y:media.location.y];
    
    InfoCallout* callout = [[InfoCallout alloc] initWithMapControl:sMap.smMapWC.mapControl BackgroundColor:nil Alignment:CALLOUT_BOTTOM];
    callout.width = 50;
    callout.height = 50;
    callout.userInteractionEnabled = YES;
    NSString* imagePath = [NSString stringWithFormat:@"%@/%@", [NSHomeDirectory() stringByAppendingString:@"/Documents"], media.paths[0]];
    
    NSString* extension = [[imagePath pathExtension] lowercaseString];
    UIImage* img;
    
    if ([extension isEqualToString:@"mp4"] || [extension isEqualToString:@"mov"]) {
        img = [MediaUtil getScreenShotImageFromVideoPath:imagePath];
    } else {
        img = [UIImage imageWithContentsOfFile:imagePath];
        // TODO 压缩图片
    }
    
    UIImageView* image = [[UIImageView alloc]initWithImage:img];
    image.frame = CGRectMake(0, 0, 50, 50);
    [callout addSubview:image];
    
    callout.mediaName = media.fileName;
    callout.mediaFilePaths = media.paths;
    callout.layerName = layerName;
    callout.httpAddress = @"";
    callout.description = @"";
    NSDate* date = [NSDate date];
    
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"yyyy-MM-dd HH:mm:ss"];
    callout.modifiedDate = [dateFormatter stringFromDate:date];
    
    callout.geoID = ((NSNumber *)[recordset getFieldValueWithString:@"SmID"]).intValue;
    
    if (gesture) {
        [callout addGestureRecognizer:gesture];
    }
    
    return callout;
}

+ (InfoCallout *)createCalloutByMedia:(SMMedia *)media recordset:(Recordset *)recordset layerName:(NSString *)layerName segesturelector:(UITapGestureRecognizer *)gesture {
    SMap* sMap = [SMap singletonInstance];
    Point2D* pt = [[Point2D alloc] initWithX:media.location.x Y:media.location.y];
    
    InfoCallout* callout = [[InfoCallout alloc] initWithMapControl:sMap.smMapWC.mapControl BackgroundColor:nil Alignment:CALLOUT_BOTTOM];
    callout.width = 50;
    callout.height = 50;
    callout.userInteractionEnabled = YES;
    NSString* imagePath = [NSString stringWithFormat:@"%@/%@", [NSHomeDirectory() stringByAppendingString:@"/Documents"], media.paths[0]];
    
    NSString* extension = [[imagePath pathExtension] lowercaseString];
    UIImage* img;
    
    if ([extension isEqualToString:@"mp4"] || [extension isEqualToString:@"mov"]) {
        img = [MediaUtil getScreenShotImageFromVideoPath:imagePath];
    } else {
        img = [UIImage imageWithContentsOfFile:imagePath];
        // TODO 压缩图片
    }
    
    UIImageView* image = [[UIImageView alloc]initWithImage:img];
    image.frame = CGRectMake(0, 0, 50, 50);
    [callout addSubview:image];
    
    callout.mediaName = media.fileName;
    callout.mediaFilePaths = media.paths;
    callout.layerName = layerName;
    callout.httpAddress = media.httpAddress;
    callout.description = media.Description;
    NSDate* date = [NSDate date];
    
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"yyyy-MM-dd HH:mm:ss"];
    callout.modifiedDate = [dateFormatter stringFromDate:date];
    
    callout.geoID = ((NSNumber *)[recordset getFieldValueWithString:@"SmID"]).intValue;
    
    [recordset edit];
    
    NSMutableString* paths = [NSMutableString string];
    for (int i = 0; i < callout.mediaFilePaths.count; i++) {
        [paths appendString:callout.mediaFilePaths[i]];
        if (i < callout.mediaFilePaths.count - 1) {
            [paths appendString:@","];
        }
    }
    
    [recordset setFieldValueWithString:@"ModifiedDate" Obj:callout.modifiedDate];
    [recordset setFieldValueWithString:@"MediaFilePaths" Obj:paths];
    [recordset setFieldValueWithString:@"Description" Obj:callout.description];
    [recordset setFieldValueWithString:@"HttpAddress" Obj:callout.httpAddress];
    
    [recordset update];
    
    if (gesture) {
        [callout addGestureRecognizer:gesture];
    }
    
    [callout setLocationX:pt.x LocationY:pt.y];
    
    return callout;
}

+ (void)addCalloutByMedia:(SMMedia *)media recordset:(Recordset *)rs layerName:(NSString *)layerName segesturelector:(UITapGestureRecognizer *)gesture {
    
//    double longitude = [rs.geometry getInnerPoint].x;
//    double latitude =  [rs.geometry getInnerPoint].y;
    
    Point2D* pt = [[Point2D alloc]initWithX:media.location.x Y:media.location.y];
//    if ([[SMap singletonInstance].smMapWC.mapControl.map.prjCoordSys type] != PCST_EARTH_LONGITUDE_LATITUDE) {//若投影坐标不是经纬度坐标则进行转换
//        Point2Ds *points = [[Point2Ds alloc]init];
//        [points add:pt];
//        PrjCoordSys *srcPrjCoorSys = [[PrjCoordSys alloc]init];
//        [srcPrjCoorSys setType:PCST_EARTH_LONGITUDE_LATITUDE];
//        CoordSysTransParameter *param = [[CoordSysTransParameter alloc]init];
//
//        //根据源投影坐标系与目标投影坐标系对坐标点串进行投影转换，结果将直接改变源坐标点串
//        [CoordSysTranslator convert:points PrjCoordSys:srcPrjCoorSys PrjCoordSys:[[SMap singletonInstance].smMapWC.mapControl.map prjCoordSys] CoordSysTransParameter:param CoordSysTransMethod:(CoordSysTransMethod)9603];
//        pt = [points getItem:0];
//    }
    
    NSString* imgPath = [NSString stringWithFormat:@"%@/%@", [NSHomeDirectory() stringByAppendingString:@"/Documents"], media.paths[0]];
//    InfoCallout* callout = [SMLayer addCallOutWithLongitude:longitude latitude:latitude image:imgPath];
    InfoCallout* callout = [SMLayer addCallOutWithLongitude:pt.x latitude:pt.y image:imgPath];
    callout.mediaName = media.fileName;
    callout.mediaFilePaths = media.paths;
    //            callout.type = media.mediaType;
    callout.layerName = layerName;
    callout.httpAddress = media.httpAddress;
    callout.description = media.Description;
    NSDate* date = [NSDate date];
    
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"yyyy-MM-dd HH:mm:ss"];
    callout.modifiedDate = [dateFormatter stringFromDate:date];
    
    callout.geoID = ((NSNumber *)[rs getFieldValueWithString:@"SmID"]).intValue;
    
    [rs edit];
    
    NSMutableString* paths = [NSMutableString string];
    //            for (NSString* path in callout.mediaFilePaths) {
    //                paths appendFormat:<#(nonnull NSString *), ...#>
    //            }
    for (int i = 0; i < callout.mediaFilePaths.count; i++) {
        [paths appendString:callout.mediaFilePaths[i]];
        if (i < callout.mediaFilePaths.count - 1) {
            [paths appendString:@","];
        }
    }
    
    [rs setFieldValueWithString:@"ModifiedDate" Obj:callout.modifiedDate];
    [rs setFieldValueWithString:@"MediaFilePaths" Obj:paths];
    [rs setFieldValueWithString:@"Description" Obj:callout.description];
    [rs setFieldValueWithString:@"HttpAddress" Obj:callout.httpAddress];
    
    [rs update];
    
    if (gesture) {
        [callout addGestureRecognizer:gesture];
    }
}

+ (void)addLineByMedias:(NSArray *)medias dataset:(DatasetVector *)dataset {
    Point2Ds* mPoints = [[Point2Ds alloc] init];
    for (SMMedia* media in medias) {
        Point2D* pt = [[Point2D alloc] initWithX:media.location.x Y:media.location.y];
        [mPoints add:pt];
    }
//    if ([[SMap singletonInstance].smMapWC.mapControl.map.prjCoordSys type] != PCST_EARTH_LONGITUDE_LATITUDE) {//若投影坐标不是经纬度坐标则进行转换
//        PrjCoordSys *srcPrjCoorSys = [[PrjCoordSys alloc]init];
//        [srcPrjCoorSys setType:PCST_EARTH_LONGITUDE_LATITUDE];
//        CoordSysTransParameter *param = [[CoordSysTransParameter alloc]init];
//
//        //根据源投影坐标系与目标投影坐标系对坐标点串进行投影转换，结果将直接改变源坐标点串
//        [CoordSysTranslator convert:mPoints PrjCoordSys:srcPrjCoorSys PrjCoordSys:[[SMap singletonInstance].smMapWC.mapControl.map prjCoordSys] CoordSysTransParameter:param CoordSysTransMethod:(CoordSysTransMethod)9603];
//    }
    
    GeoLine* mLine = [[GeoLine alloc] initWithPoint2Ds:mPoints];
    GeoStyle* style = [[GeoStyle alloc] init];
    [style setMarkerSize:[[Size2D alloc] initWithWidth:2 Height:2]];
    [style setLineColor:[[Color alloc] initWithR:70 G:128 B:223]];
    [style setLineWidth:1];
    [style setMarkerSymbolID:351];
    [style setFillForeColor:[[Color alloc] initWithR:70 G:128 B:223]];
    [mLine setStyle:style];
    
    Recordset* mRecordset = [dataset recordset:NO cursorType:DYNAMIC];
    
    [mRecordset addNew:mLine];
    
    [mRecordset moveLast];
    if ([mRecordset edit]) {
        [mRecordset setStringWithName:@"MediaName" StringValue:@"TourLine"];
    }
    
    [mRecordset update];
    
    [mRecordset dispose];
}

@end
